src/censored.jl

```diff
@@ -0,0 +1,407 @@
+# 𝔼_{x ~ τ}[-log d(x)] = S[τ] - log P_{x ~ d₀}(l ≤ x ≤ u)
+#   + (P_{x ~ d₀}(x = l) (log P_{x ~ d₀}(x = l) - log P_{x ~ d₀}(x ≤ l)) + 
+#      P_{x ~ d₀}(x = u) (log P_{x ~ d₀}(x = u) - log P_{x ~ d₀}(x ≥ u))
+#   ) / P_{x ~ d₀}(l ≤ x ≤ u),
+# where S[τ] is the entropy of τ.
+
+function entropy(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    if lower !== missing
+        prob_lower_inc = cdf(d0, lower)
+        pl = value_support(typeof(d0)) === Discrete ? pdf(d0, lower) : 0
+        prob_lower = prob_lower_inc - pl
+        entropy_lower = -xlogx(prob_lower_inc)
+    else
+        pl = prob_lower = entropy_lower = 0
+    end
+    if upper !== missing
+        prob_upper = ccdf(d0, upper)
+        pu = value_support(typeof(d0)) === Discrete ? pdf(d0, upper) : 0
+        entropy_upper = -xlogx(prob_upper + pu)
+    else
+        pu = prob_upper = entropy_upper = 0
+    end
+    result = entropy_lower + entropy_upper
+    prob_interval = 1 - (prob_lower + prob_upper)
+    # truncation contains no probability
+    iszero(prob_interval) && return result
+
+    dtrunc = _to_truncated(d)
+    result += prob_interval * entropy(dtrunc) - xlogx(prob_interval) + xlogx(pl) + xlogx(pu)
+    return result
+end
+
+
+#### Evaluation
+
+function pdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if lower !== missing && x == lower
+        result = cdf(d0, x)
+        _eqnotmissing(x, upper) ? one(result) : result
+    elseif _eqnotmissing(x, upper)
+        _ccdf_inclusive(d0, x)
+    else
+        result = pdf(d0, x)
+        _in_open_interval(x, lower, upper) ? result : zero(result)
+    end
+end
+
+function logpdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return if lower !== missing && x == lower
+        result = logcdf(d0, x)
+        _eqnotmissing(x, upper) ? zero(result) : result
+    elseif _eqnotmissing(x, upper)
+        _logccdf_inclusive(d0, x)
+    else
+        result = logpdf(d0, x)
+        _in_open_interval(x, lower, upper) ? result : oftype(result, -Inf)
+    end
+
+end
+
+function loglikelihood(d::Censored, x::AbstractArray{<:Real})
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    one_like_x = one(first(x))
+    if lower === missing
+        T = float(Base.promote_eltype(upper, one_like_x))
+    elseif upper === missing
+        T = float(Base.promote_eltype(lower, one_like_x))
+    else
+        T = float(Base.promote_eltype(lower, upper, one_like_x))
+    end
+    log_prob_lower = lower === missing ? 0 : logcdf(d0, T(lower))
+    log_prob_upper = upper === missing ? 0 : _logccdf_inclusive(d0, T(upper))
+    logzero = T(-Inf)
+
+    return sum(x) do xi
+        _in_open_interval(xi, lower, upper) && return logpdf(d0, xi)
+        _eqnotmissing(xi, lower) && return log_prob_lower
+        _eqnotmissing(xi, upper) && return log_prob_upper
+        return logzero
+    end
```

Comment:
I think it's still not type stable: Similar to before we can take `lower = 0.0`, `upper = 10.0`, `d0 = Normal{Int}(0, 1)`, and `x = [1.0f0, -2.0f0]`. Then `T = Float64` and hence `log_prob_lower isa Float64`, `log_prob_upper isa Float64`, and `logzero isa Float64`. The first observation `x[1]` yields `logpdf(d0, x[1]) isa Float32` whereas the second observation `x[2]` returns `logzero isa Float64`.

In general, it seems that we should not return something of type `T` since this makes `loglikelihood` inconsistent with `logpdf`: if `x` is a vector of `Float32`, then the `logpdf` is also a `Float32` even if the bounds are a `Float64` (which is the correct IMO) but this implementation of the `loglikelihood` would return a `Float64` (at least sometimes).

Reply:
I'm seeing something different from you:
```julia
julia> d = censored(Normal{Int}(0, 1), 0.0, 10.0)
Censored(Normal{Int64}(μ=0, σ=1), range=(0.0, 10.0))

julia> x = [1.0f0, -2.0f0]
2-element Vector{Float32}:
  1.0
 -2.0

julia> @inferred loglikelihood(d, x)
-Inf

julia> @inferred logpdf(d, x[1])
ERROR: return type Float64 does not match inferred return type Union{Float32, Float64}
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] top-level scope
   @ REPL[32]:1

julia> @inferred logpdf(d, x[2])
ERROR: return type Float64 does not match inferred return type Union{Float32, Float64}
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] top-level scope
   @ REPL[33]:1
```
So `loglikelihood` is more type-stable than `logpdf`. Incidentally, `logpdf` is type-unstable here because `logcdf` for `Normal` is type-unstable (see `result` below):
```julia
julia> @code_warntype logpdf(d, x[1])
MethodInstance for Distributions.logpdf(::Distributions.Censored{Normal{Int64}, Continuous, Float64, Float64, Float64}, ::Float32)
  from logpdf(d::Distributions.Censored, x::Real) in Distributions at /home/sethaxen/projects/GitHub/Distributions.jl/src/censored.jl:337
Arguments
  #self#::Core.Const(Distributions.logpdf)
  d::Distributions.Censored{Normal{Int64}, Continuous, Float64, Float64, Float64}
  x::Float32
Locals
  result::Union{Float32, Float64}
  upper::Float64
  lower::Float64
  d0::Normal{Int64}
Body::Union{Float32, Float64}
1 ──       Core.NewvarNode(:(result))
│          (d0 = Base.getproperty(d, :uncensored))
│          (lower = Base.getproperty(d, :lower))
│          (upper = Base.getproperty(d, :upper))
│    %5  = (lower !== Distributions.missing)::Core.Const(true)
└───       goto #6 if not %5
2 ── %7  = (x == lower)::Bool
└───       goto #6 if not %7
3 ──       (result = Distributions.logcdf(d0, x))
│    %10 = Distributions._eqnotmissing(x, upper)::Bool
└───       goto #5 if not %10
4 ── %12 = Distributions.zero(result)::Union{Float32, Float64}
└───       return %12
5 ──       return result
6 ┄─ %15 = Distributions._eqnotmissing(x, upper)::Bool
└───       goto #8 if not %15
7 ── %17 = Distributions._logccdf_inclusive(d0, x)::Union{Float32, Float64}
└───       return %17
8 ──       (result = Distributions.logpdf(d0, x))
│    %20 = Distributions._in_open_interval(x, lower, upper)::Bool
└───       goto #10 if not %20
9 ──       return result::Float64
10 ─ %23 = result::Float64
│    %24 = -Distributions.Inf::Core.Const(-Inf)
│    %25 = Distributions.oftype(%23, %24)::Core.Const(-Inf)
└───       return %25
```

Reply:
If the support of `d0` is `1:10`, but `lower=2.5f0`, this adds a `Float32` to the support of `d`, and for type-stability, this should be reflected in all outputs. So I do think we need to take into account the types of `lower` and `upper` in the return values e.g. `logpdf`, and that should probably be checked.

Reply:
The latest commits take type-stability very seriously and test it thoroughly.